[AST] Start using conformance access paths in SubstitutionMap

DougGregor opened the pull request:
Reimplement `SubstitutionMap::lookupConformance()` in terms of the `ConformanceAccessPath`, eliminating the ugly "parent map" hack and ensuring that all stages of substitutions are depending on the same core abstractions (`GenericSignature`, `ConformanceAccessPath`.

While here, remove the error blocking the use of associated type where clauses [SE-0142](https://github.com/apple/swift-evolution/blob/master/proposals/0142-associated-types-constraints.md), because we should be able to start experimenting with them in earnest now regardless of language mode.

Fixes rdar://problem/31041997 and rdar://problem/31309863.

DougGregor committed:
[AST] Track the generic signature/environment for a substitution map.

Substitution maps are effectively tied to a particular generic
signature or environment; keep track of that signature/environment so
that we can (eventually) use it to find conformances.

DougGregor committed:
Check conformances against the requirement signature using module lookup.

When we are checking the conformance of a nominal type to a particular
protocol, we check each of the requirements in the requirement
signature, substituting in our Self type and (by extension) type
witnesses. However, when we're trying to fulfill requirements for
associated types (which might be mapped to concrete types), perform
the conformance lookup in the module of the conformance itself---not
the SubstitutionMap, which can't possibly have them at this point.

Fixes rdar://problem/31041997.

DougGregor committed:
[SubstitutionMap] Use conformance access paths to find conformances.

When looking up a particular conformance in a SubstitutionMap, use the
associated generic signature to determine how to find that conformance
(via a conformance access path). Follow that conformance access path
to retrieve the desired conformance.

FIXME: There is still one failure, here

  Constraints/generic_overload.swift

which appears to be because we either have the wrong generic signature
for the override case or we are looking in the wrong map.

DougGregor committed:
[SubstitutionMap] Eliminate the parent map and handle archetype conformances.

When looking for a conformance for an archetype, map it out of context
to compute the conformance access path, then do the actual lookups
based on mapping the starting type back into the context. Eliminate
the parent map and "walk the conformances" functionality.

huonw commented on lib/AST/SubstitutionMap.cpp:
```diff
@@ -64,173 +72,93 @@ Type SubstitutionMap::lookupSubstitution(CanSubstitutableType type) const {
+SubstitutionMap::lookupConformance(CanType type, ProtocolDecl *proto) const {
+  // If we have an archetype, map out of the context so we can compute a
+  // conformance access path.
+  GenericEnvironment *genericEnv = nullptr;;
```
Double `;`.

DougGregor replied:
Got it, thanks

huonw commented on test/Generics/protocol_where_clause.swift:
```diff
@@ -65,3 +65,12 @@ struct BadConcreteNestedConforms: NestedConforms {
+protocol P1 {}
+protocol P2 { associatedtype T }
+protocol P3 { associatedtype U }
+protocol P4: P3 where U: P2, U.T: P1 {}
```
Hm, isn't this isomorphic to the `NestedConforms` test above? @slavapestov's targeted fix https://github.com/apple/swift/pull/8428 that made that work includes a "rip this out once conformance access paths are plumbed through" comment, can that be executed now?

DougGregor replied:
Oh, you're right. Well then! I'll see if I can rip out the hack in #8428 now; it might require better handling of concrete types in `GenericSignature`.

DougGregor committed:
[SubstitutionMap] Cope with missing conformances in lookupConformance().

When asking a substitution map for a conformance, it's okay if the
conformance isn't there---just detect this case and return None.

Also, collapse a redundant testcase Huon noted.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor committed:
Enable associated type where clauses in all language modes

We're ready to start experimenting with and rolling out this feature
in earnest.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test Linux

DougGregor commented:
@swift-ci please smoke test Linux

DougGregor committed:
[GSB] Properly compute archetype anchor during same-type DFS.

This would cause spurious idempotency failures with generic signature canonicalization.

DougGregor committed:
[SubstitutionMap] Make conformance lookup robust against ill-formed conformances.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge

slavapestov commented on lib/AST/SubstitutionMap.cpp:
```diff
@@ -64,173 +72,115 @@ Type SubstitutionMap::lookupSubstitution(CanSubstitutableType type) const {
+    [&](Type type, ProtocolDecl *proto) -> Optional<ProtocolConformanceRef> {
+      // We're working relative to a generic environment, map into that
+      // context before looking into the conformance map.
+      if (genericEnv)
```
It seems silly to map the type out of context only to map it back in again. What do you think of always using interface types as keys in the conformance map, even for generic environment-derived substitutions?

DougGregor replied:
I think it's a great idea. What I'd like to do is turn the conformance map into a flat array, arranged according to the protocol requirements in the `GenericSignature`, the way we do with `NormalProtocolConformance` and its "signature conformances".

slavapestov commented on lib/AST/SubstitutionMap.cpp:
```diff
@@ -64,173 +72,115 @@ Type SubstitutionMap::lookupSubstitution(CanSubstitutableType type) const {
-  });
-}
+  auto canonType = genericSig->getCanonicalTypeInContext(type, mod);
+  auto accessPath =
```
It seems expensive to go through the canonical GSB to do getCanonicalTypeInContext(), and then feed the result back into the GSB for getConformanceAccessPath(). Can getConformanceAccessPath() instead do the right thing with non-canonical types?

DougGregor replied:
Actually, `getConformanceAccessPath()` can work on non-canonical types, because it implicitly canonicalizes when looking up the potential archetype. I think the canonicalization you refer to is unnecessary.

slavapestov commented on lib/AST/SubstitutionMap.cpp:
```diff
@@ -64,173 +72,115 @@ Type SubstitutionMap::lookupSubstitution(CanSubstitutableType type) const {
-        return recordOrReturn(conformance.getInherited(proto));
+    // If we've hit an abstract conformance, everything from here on out is
+    // abstract.
+    // FIXME: This may not always be true, but it holds for now.
```
This is of course the infamous SR-3500

slavapestov commented:
Looks cool! I'm mostly concerned with making lookupConformance() cheaper. Anything we can do to eliminate redundant canonicalization steps or to memoize more stuff would be good.

DougGregor committed:
Handle substitution in InheritedProtocolConformance::getAssociatedConformance.

Fixes a regression introduced when we switched SubstitutionMap::lookupConformance() to conformance access paths.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
That last commit should address the corelibs Foundation failure.

swiftix commented:
@DougGregor Have you tried to perform a build with these changes and enabling the SpecializeGenericSubstitutions in Generics.cpp? It may result in exposing more edge cases which are not addressed by this PR yet.
